Makefile

```diff
@@ -1,58 +1,75 @@
+GO_CMD        = go
 WORKDIR      := $(shell pwd)
-NATIVEOS     := $(shell go version | awk -F '[ /]' '{print $$4}')
-NATIVEARCH   := $(shell go version | awk -F '[ /]' '{print $$5}')
-GO_PKGS      := $(shell go list ./... | grep -v -e "/vendor/" -e "/example")
+NATIVEOS     := $(shell $(GO_CMD) version | awk -F '[ /]' '{print $$4}')
+NATIVEARCH   := $(shell $(GO_CMD) version | awk -F '[ /]' '{print $$5}')
+GO_PKGS      := $(shell $(GO_CMD) list ./... | grep -v -e "/vendor/" -e "/example")
 GO_FILES     := $(shell find cmd -type f -name "*.go")
+GO_TOOLS      = github.com/axw/gocov/gocov github.com/AlekSi/gocov-xml
+
+BIN_DIR            = $(WORKDIR)/bin
+TARGET             = target
+TARGET_DIR         = $(WORKDIR)/$(TARGET)
+INTEGRATION       := vsphere
+SHORT_INTEGRATION := vsphere
+BINARY_NAME        = nri-$(INTEGRATION)
+CONTAINER_IMAGE    = $(PROJECT_NAME)-builder
+CONTAINER          = $(PROJECT_NAME)
+CONTAINER_PATH     = /go/src/$(PROJECT_NAME)
+
+LINTER         = golangci-lint
+LINTER_VERSION = 1.27.0
+SNYK_BIN       = snyk-linux
+SNYK_VERSION   = v1.361.3
 
-GO_CMD        = go
-GOLINTER      = golangci-lint
+all: build
+build-local: clean compile test
+build: build-container-image delete-container test-container delete-container
 
-BIN_DIR    = $(WORKDIR)/bin
-TARGET = target
-TARGET_DIR       = $(WORKDIR)/$(TARGET)
-INTEGRATION  := vsphere
-SHORT_INTEGRATION  := vsphere
-BINARY_NAME   = nri-$(INTEGRATION)
+build-container-image:
+	@docker build --no-cache -t $(CONTAINER_IMAGE) .
 
-GOTOOLS       = github.com/kardianos/govendor  github.com/axw/gocov/gocov github.com/AlekSi/gocov-xml
+test-container:
+	@echo "make test" | docker run --name $(CONTAINER) -i $(CONTAINER_IMAGE)
+	@docker cp $(CONTAINER):$(CONTAINER_PATH)/coverage.xml .
 
-SNYK_VERSION  = v1.361.3
-SNYK_BIN = snyk-linux
+compile-container: bin
+	@echo "make compile" | docker run --name $(CONTAINER) -i $(CONTAINER_IMAGE)
+	@docker cp $(CONTAINER):$(CONTAINER_PATH)/bin/$(BINARY_NAME) $(BINS_DIR)
 
-all: build
-build: clean test compile
+delete-container:
+	-docker rm -f $(CONTAINER) 2>/dev/null
```

Comment:
shouldn't this be "@docker rm ..."  ?

Reply:
wit the - before the command, makefile won't fail in case the command exits with an error code (we try delete the container before we run it)